Update technical design (#495)
kyma-project/istio: docs/contributor/04-10-technical-design.md

have a better extensibility and maintainability. This means each of these components must have its clearly separated responsibility and must work in isolation when assessing whether reconciliation is required, applying changes, and returning a status.
To understand the reasons for the technical design of the Kyma Istio Operator, refer to the [Architecture Decision Record](https://github.com/kyma-project/istio/issues/135).

The following diagram illustrates the Kyma Istio Operator and its components:

![Kyma Istio Operator Overview](../assets/istio-operator-overview.svg)

## Istio CR

The [Istio CR](../user/03-technical-reference/istio-custom-resource/01-30-istio-custom-resource.md) a namespace-scoped resource that is used to manage the Istio installation. The reason for Istio CR being namespace-scoped is that it was created and used before the Kyma Istio Operator was introduced.
There is no advantage to the Istio CR being namespace-scoped as the Kyma Istio Operator only supports one Istio CR per cluster. However, it is not possible to change it to cluster-scoped without breaking changes.

> If multiple Istio CRs exist on the cluster, the Kyma Istio operator will only use the oldest Istio CR for reconciliation and put the other Istio CRs in an error state.

Reviewer: Should we also mention that it needs to be in `kyma-system` and the recommended name is `default`?

Author: Hmm, good point. I will think about what's the best way to add it to the design, because the defaults should also be mentioned in a more customer facing documentation and I can't find this. So I will check where we need to add this as well.

Author: https://github.com/kyma-project/istio/pull/495/commits/8ca3eefd59acb3167b8ccd8150a020780c0ec3ea

Author: I added it: https://github.com/kyma-project/istio/pull/495/commits/8ca3eefd59acb3167b8ccd8150a020780c0ec3ea

## Istio version

The version of Istio is coupled to the version of the Kyma Istio Operator. This means that a particular version of the operator is responsible for a particular version of Istio and a new version of the operator is released to support a new version of Istio.
Upgrading to a new version of the operator will automatically update Istio if the version of Istio has changed in the new operator version.

## Version upgrade

The Istio version upgrade is limited to supporting only one minor version (1.2.3 -> 1.3.0). This implies that this is also relevant when upgrading the Kyma Istio Operator to a new version.
This means that when upgrading the operator to a new version, a minor version of the operator can only be skipped if the difference between the Istio versions in the new operator version is no more than one minor version.

If the difference between the current and the target version of Istio is greater than one minor version (1.2.3 -> 1.4.0), the reconciliation will fail.

> The same applies to major version upgrades (1.2.3 -> 2.0.0) and downgrades (1.2.3 -> 1.2.2).

Reviewer: Downgrades are not supported

Author: Yes, that's what is mentioned in the docs. That in case of major upgrade or downgrade, the reconciliation will fail as well. Maybe it's not obvious enough and we should write it like that:
```suggestion
The reconciliation will also fail for major version upgrades (1.2.3 -> 2.0.0) and downgrades (1.2.3 -> 1.2.2), as these are also not supported.
```

Reviewer: I just don't see if mentioning the path "(1.2.3 -> 1.2.2)" makes sense, as it doesn't apply to either patch or minor, it's a general behaviour. At least specifying that it applies to `any` downgrade maybe would be better.

Author: Ah Okay, now I get your point.
So we should rather have it like that?
```suggestion
The reconciliation will also fail for major version upgrades (1.2.3 -> 2.0.0) and version downgrades, as these are also not supported.
```

Reviewer: 👍

Upgrading to a new version of the operator will automatically update Istio if the version of Istio has changed in the new operator version.

## Version upgrade

The Istio version upgrade is limited to supporting only one minor version (1.2.3 -> 1.3.0). This implies that this is also relevant when upgrading the Kyma Istio Operator to a new version.
This means that when upgrading the operator to a new version, a minor version of the operator can only be skipped if the difference between the Istio versions in the new operator version is no more than one minor version.

If the difference between the current and the target version of Istio is greater than one minor version (1.2.3 -> 1.4.0), the reconciliation will fail.
The reconciliation will also fail for major version upgrades (1.2.3 -> 2.0.0) and version downgrades, as these are also not supported.

## Istio Controller

The Istio Controller is a [Kubernetes controller](https://kubernetes.io/docs/concepts/architecture/controller/), which is implemented using the [Kubebuilder](https://book.kubebuilder.io/) framework.

> The controller is responsible for handling the [Istio CR](../user/03-technical-reference/istio-custom-resource/01-30-istio-custom-resource.md). The controller reconciles only Istio CRs in the kyma-system namespace.

Author: 👍 I added the missing `in`in your suggestion.
```suggestion
Istio Controller manages and reconciles Istio CRs, which are in the `kyma-system` namespace.
```

and to decide whether a certain configuration has been changed that results in a certain action, for example restarting the Ingress Gateway.

#### Istio Ingress Gateway restart
The component also detects changes in the `numTrustedProxies` configuration and restarts the Istio Ingress Gateway accordingly. 
Whenever a change in the `numTrustedProxies` configuration is detected, the Pods in the `istio-system/istio-ingressgateway` deployment are restarted.

### Istio ResourcesReconciliation

Istio ResourcesReconciliation is a component responsible for applying resources dependent on Istio, such as VirtualService or EnvoyFilter, and ensuring that the state of the Istio service mesh is configured correctly based on those resources.
To maintain the correct state, the component provides [Restart Predicates](#restart-predicates) on a per-resource basis, which are consumed by the [IngressGatewayReconciler](#ingressgatewayreconciler) and [ProxySidecarReconciliation](#proxysidecarreconciliation) components.
To understand the reasons for the predicates, refer to the [Architecture Decision Record](https://github.com/kyma-project/istio/issues/278).

> For resources that are reconciled by this component, which are not Istio resources and would therefore remain on the cluster if Istio is uninstalled, the OwnerReference is set to the Istio CR. 

Reviewer 2: ```suggestion
When the Istio InstallationReconciliation component reconciles resources that are not Istio resources, it sets their OwnerReference to the Istio CR.
```

> This means that if the Istio CR is deleted, these resources are also deleted.

Reviewer 2: ```suggestion
As a result, if you remove the Istio CR, any associated resources that would normally remain on the cluster are also deleted.
```

### ProxySidecarReconciliation

> ProxySidecarReconciliation component is responsible for keeping the proxy sidecars in the desired state. It restarts Pods that are part of Service Mesh or

Reviewer 2: ```suggestion
The ProxySidecarReconciliation component is responsible for keeping the proxy sidecars in the desired state. It restarts Pods that are part of the service mesh or
```

> that need to be added to the Service Mesh.

Reviewer 2: ```suggestion
that must be added to the service mesh.
```

> The desired state is represented by [Istio CR](../user/03-technical-reference/istio-custom-resource/01-30-istio-custom-resource.md) and [Istio Version](#istio-version).

Reviewer 2: ```suggestion
The Istio CR and [Istio Version](#istio-version) represent the desired state.
```

> The following triggers for a restart are be covered by this component:

Reviewer 2: ```suggestion
This component covers the following restart triggers:
```

- Restart Pods with proxy sidecar when CNI config changes.
- Restart Pods with proxy sidecar after an Istio version update.
- Restart Pods with proxy sidecar when proxy resources change.

> - Restart Pods if they match [Restart Predicates](#restart-predicates) that the [Istio ResourcesReconciliation component](#istio-resourcesreconciliation) specifies (for example, being up-to-date with EnvoyFilter)

Reviewer 2: ```suggestion
- Restart Pods if they match [Restart Predicates](#restart-predicates) that the [Istio ResourcesReconciliation component](#istio-resourcesreconciliation) specifies (for example, being up-to-date with EnvoyFilter).
```

### IngressGatewayReconciler

> The IngressGatewayReconciler component is responsible for restarting Istio Ingress Gateway. The component consumes a list of [Restart Predicates](#restart-predicates) that specify when the Ingress Gateway should be restarted.

Reviewer 2: ```suggestion
IngressGatewayReconciler is responsible for restarting Istio Ingress Gateway. The component consumes a list of [Restart Predicates](#restart-predicates) that determine when the restart should occur.
```

> To understand the reasons for the predicates, refer to the [Architecture Decision Record](https://github.com/kyma-project/istio/issues/278).

Reviewer 2: ```suggestion
```

## Restart Predicates

> The restart predicates are used by the [IngressGatewayReconciler](#ingressgatewayreconciler) and [ProxySidecarReconciliation](#proxysidecarreconciliation) components.

Reviewer 2: ```suggestion
The [IngressGatewayReconciler](#ingressgatewayreconciler) and [ProxySidecarReconciliation](#proxysidecarreconciliation) components use Restart Predicates.
```

> A predicate specifies when the aforementioned components should be restarted. Depending on the implemented interfaces, a predicate can either trigger a restart of Ingress Gateways, Proxy Sidecars or Ingress Gateways and Proxy Sidecars.

Reviewer 2: ```suggestion
Depending on the implemented interfaces, a predicate can trigger a restart of Ingress Gateways, Proxy Sidecars, or both Ingress Gateways and Proxy Sidecars.
```

> For cases where it isn't trivial to check whether the configuration has been applied to the cluster state, a timestamp-based approach is used. For example, the `envoy_filter_allow_partial_referer` resource is annotated with the `istios.operator.kyma-project.io/updatedAt` annotation, which includes the timestamp of its last update.

Reviewer 2: ```suggestion
For cases where it isn't trivial to check whether the configuration has been applied to the cluster state, Restart Predicates use a timestamp-based approach. For example, the `envoy_filter_allow_partial_referer` resource has the `istios.operator.kyma-project.io/updatedAt` annotation, which includes the timestamp of its last update.
```

> The predicate initiates a restart of the sidecar and Ingress Gateway if the target had been created prior to this timestamp.

Reviewer 2: ```suggestion
The predicate initiates a restart of the sidecar and Ingress Gateway if the target was created before this timestamp.
```
